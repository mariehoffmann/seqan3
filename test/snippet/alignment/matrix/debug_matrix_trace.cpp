#include <iostream>

#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/alignment/matrix/debug_matrix.hpp>
#include <seqan3/core/debug_stream.hpp>
#include <seqan3/range/view/to_char.hpp>

int main()
{
    // using namespace seqan3;
    using seqan3::detail::debug_matrix;
    using seqan3::operator""_dna4;
    using seqan3::operator|;

    std::vector<seqan3::dna4> database = "AACCGGTT"_dna4;
    std::vector<seqan3::dna4> query = "ACGT"_dna4;

    auto N = seqan3::detail::trace_directions::none;
    auto D = seqan3::detail::trace_directions::diagonal;
    auto U = seqan3::detail::trace_directions::up;
    auto L = seqan3::detail::trace_directions::left;

    seqan3::detail::row_wise_matrix trace_matrix
    {
        std::vector
        {
            N,L,L  ,L  ,L  ,L  ,L  ,L,L  ,
            U,D,D|L,L  ,L  ,L  ,L  ,L,L  ,
            U,U,D  ,D  ,D|L,L  ,L  ,L,L  ,
            U,U,D|U,D|U,D  ,D  ,D|L,L,L  ,
            U,U,D|U,D|U,D|U,D|U,D  ,D,D|L
        },
        5u,
        9u
    };

    seqan3::debug_stream << "database:\t" << database << std::endl;
    seqan3::debug_stream << "query:\t\t" << query << std::endl;
    seqan3::debug_stream << std::endl;

    seqan3::debug_stream << "trace_matrix: " << trace_matrix.cols() << " columns and "
              << trace_matrix.rows() << " rows" << std::endl;

    // Prints out the matrix in a convenient way
    seqan3::debug_stream << trace_matrix << std::endl; // without sequences
    seqan3::debug_stream << debug_matrix{trace_matrix, database, query} << std::endl; // with sequences
    seqan3::debug_stream << seqan3::fmtflags2::utf8 << debug_matrix{trace_matrix, database, query}; // as utf8
    return 0;
}
