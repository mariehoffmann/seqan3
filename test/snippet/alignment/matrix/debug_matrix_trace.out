database:	AACCGGTT
query:		ACGT

trace_matrix: 9 columns and 5 rows
 ;  ;  ;  ;  ;  ;  ;  ;  ;  ;
 ;N ;L ;L ;L ;L ;L ;L ;L ;L ;
 ;U ;D ;DL;L ;L ;L ;L ;L ;L ;
 ;U ;U ;D ;D ;DL;L ;L ;L ;L ;
 ;U ;U ;DU;DU;D ;D ;DL;L ;L ;
 ;U ;U ;DU;DU;DU;DU;D ;D ;DL;

 ;  ;A ;A ;C ;C ;G ;G ;T ;T ;
 ;N ;L ;L ;L ;L ;L ;L ;L ;L ;
A;U ;D ;DL;L ;L ;L ;L ;L ;L ;
C;U ;U ;D ;D ;DL;L ;L ;L ;L ;
G;U ;U ;DU;DU;D ;D ;DL;L ;L ;
T;U ;U ;DU;DU;DU;DU;D ;D ;DL;

 ║ε ║A ║A ║C ║C ║G ║G ║T ║T ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
ε║↺ ║← ║← ║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
A║↑ ║↖ ║↖←║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
C║↑ ║↑ ║↖ ║↖ ║↖←║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
G║↑ ║↑ ║↖↑║↖↑║↖ ║↖ ║↖←║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
T║↑ ║↑ ║↖↑║↖↑║↖↑║↖↑║↖ ║↖ ║↖←║
